### PR TITLE
[AIR] Result Restoration from Local Trial Directory

### LINKED ISSUE #34855: [Train] Get a run's best `Checkpoint`
Get an experiment run's best checkpoint based on metrics and direction. Or if a `CheckpointConfig` was used with a `checkpoint_score_attribute` (ex. val_loss) and `checkpoint_score_order` (ex. min), maybe we can just save the best `Checkpoint` by default.

[Update for Clarification]
`checkpoint` and `best_checkpoints` are great if I have access to the `Result` object (ie. just finished training/tuning)
but if I want load load the best checkpoint for inference later (no longer have a Result object), I want to be able to do:
```python
predictor = TorchPredictor.from_checkpoint(best_checkpoint)
```

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

Users reported that they are unable to retrieve trial results after we lost the in-memory Result object. This pull request enables restoration from a local trial directory.

- Restore the metric dataframe from `result.json`
- Restore all checkpoints from ckpt directory
- Restore error object from `error.pkl`
- Provide a utility function get_best_checkpoint(metric, mode).

This is a stacked PR based on #35404. The major changes are:
- [python/air/result.py](https://github.com/ray-project/ray/pull/35406/files#diff-1f833ee57e965a1954c1815317a46ccf9720beba52c23d31e74f185570388122)
- [python/air/tests/test_result.py](https://github.com/ray-project/ray/pull/35406/files#diff-75e183bbfdcb44aae03019e072e878d05477649d3e691434c50682023b733061)

## Related issue number

Closes #34855 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by GokuMohandas** [comment]
Hi @woshiyyya, I tested these updates on my end-to-end workflow and everything worked great! I was able to switch out my verbose patch (where I was parsing the progress.csv, etc.) to just one line now.

```python
# Evaluate on test split
artifact_dir = urlparse(mlflow.get_run(run_id).info.artifact_uri).path  # this is MLflow specific
results = Result.from_path(artifact_dir)
best_checkpoint = results.get_best_checkpoint(metric="val_loss", mode="min")
predictor = TorchPredictor.from_checkpoint(best_checkpoint)
...
```

File: python/ray/air/result.py [hunk 1]
```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
...
+            # TODO(air-team): make metrics a property of Checkpoint
+            best_checkpoints = list(zip(checkpoints, checkpoint_metrics))
+            latest_checkpoint = max(checkpoints, key=lambda ckpt: ckpt.id)
+        else:
+            best_checkpoints = latest_checkpoint = None
+
+        # Restore the trial error if it exists
+        error = None
+        error_file_path = Path(local_path) / EXPR_ERROR_PICKLE_FILE
+        if os.path.exists(error_file_path):
```
Comment (justinvyu) [inline python/ray/air/result.py:182]: Nit: can use pathlib

```
error_file_path.exists()
```

File: python/ray/air/result.py [hunk 2]
```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
...
+            checkpoint=latest_checkpoint,
+            _local_path=local_path,
+            _remote_path=None,
+            metrics_dataframe=metrics_df,
+            best_checkpoints=best_checkpoints,
+            error=error,
+        )
+
+    @PublicAPI(stability="alpha")
+    def get_best_checkpoint(self, metric: str, mode: str) -> Optional[Checkpoint]:
```
Comment (justinvyu) [inline python/ray/air/result.py:196]: With ResultGrid, we're able to keep track of the metric and mode passed in through the `TuneConfig`. Now, with only the information in a single trial directory, we don't currently keep that metric/mode around, so a user must pass it in. 

Should we save that info around?

File: python/ray/air/result.py [hunk 3]
```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
...
+
+        assert mode in [
+            "max",
+            "min",
+        ], f'Unsupported mode: {mode}. Please choose from ["min", "max"]!'
+
+        op = max if mode == "max" else min
+        valid_checkpoints = [
+            ckpt_info for ckpt_info in self.best_checkpoints if metric in ckpt_info[1]
+        ]
```
Comment (justinvyu) [inline python/ray/air/result.py:219]: If there are no valid checkpoints, I think we should throw a warning and give the users the set of metric names that can be used. Then, switch this method to always return type to non-optional.

File: python/ray/air/result.py [hunk 4]
```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
...
+            _local_path=local_path,
+            _remote_path=None,
+            metrics_dataframe=metrics_df,
+            best_checkpoints=best_checkpoints,
+            error=error,
+        )
+
+    @PublicAPI(stability="alpha")
+    def get_best_checkpoint(self, metric: str, mode: str) -> Optional[Checkpoint]:
+        """Gets best persistent checkpoint of this trial.
```
Comment (justinvyu) [inline python/ray/air/result.py:197]: Let's not mention `persistent` in this user-facing docstring. User shouldn't have to worry about these distinctions.

File: python/ray/air/tests/test_result.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+@pytest.fixture
+def ray_start_4_cpus():
+    address_info = ray.init(num_cpus=4)
+    yield address_info
+    # The code after the yield will run as teardown code.
+    ray.shutdown()
+
+
+def test_result_restore(ray_start_4_cpus):
```
Comment (justinvyu) [inline python/ray/air/tests/test_result.py:19]: Can we test it for a Train run and a Tune run?

**Review by justinvyu — COMMENTED** [review]
Looks good so far.

Is there a stacked PR for the cloud path case? If not, let's just add a disclaimer in the docstring about only local path being supported for now.

File: python/ray/air/result.py [hunk 6]
```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
...
+        Returns:
+            :class:`Checkpoint <ray.air.Checkpoint>` object, or None if there is
+            no valid checkpoint associated with the metric.
+        """
+        assert self.best_checkpoints, "No checkpoint exists in the trial directory!"
+
+        assert mode in [
+            "max",
+            "min",
+        ], f'Unsupported mode: {mode}. Please choose from ["min", "max"]!'
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/result.py:214]: Thanks Justin! That's a really great point!

File: python/ray/air/checkpoint.py [hunk 7]
```diff
@@ -856,6 +856,16 @@ def _get_checkpoint_type(
...
 
+    @property
+    def id(self) -> int:
+        """Return the 0-based id of the checkpoint."""
+        return getattr(self, CHECKPOINT_ID_ATTR)
+
+    @property
+    def iteration(self) -> int:
+        """Return the 1-based iteration number of the checkpoint."""
+        return getattr(self, CHECKPOINT_ID_ATTR) + 1
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/checkpoint.py:867]: But we still need to know the checkpoint id, in order to align with the metrics. 
Do you think it's better to parse the checkpoint folder name to get that?

File: python/ray/air/result.py [hunk 8]
```diff
@@ -48,6 +54,7 @@ class Result:
     _local_path: Optional[str] = None
     _remote_path: Optional[str] = None
     _items_to_repr = ["error", "metrics", "path", "checkpoint"]
+    _restore_required_files = [EXPR_RESULT_FILE, EXPR_PROGRESS_FILE]
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/result.py:57]: Currently we are using the `result.json`. I can keep one here. 
p.s. are we trying to deprecate `progress.csv` now?

File: python/ray/air/result.py [hunk 9]
```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
...
+
+    @staticmethod
+    def _validate_trial_dir(trial_dir: str):
+        """Check the validity of the local trial folder."""
+
+        assert os.path.exists(trial_dir), f"Trail folder {trial_dir} doesn't exists!"
+
+        all_files = os.listdir(trial_dir)
+        for file in Result._restore_required_files:
+            assert file in all_files, f"{file} not found in the trial folder!"
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/result.py:129]: OK

File: python/ray/air/result.py [hunk 10]
```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
...
+            json_list = [json.loads(line) for line in f if line]
+            metrics_df = pd.json_normalize(json_list, sep="/")
+
+        metrics = metrics_df.iloc[-1].to_dict() if not metrics_df.empty else {}
+
+        # Restore all checkpoints from the checkpoint folders
+        ckpt_dirs = [
+            os.path.join(local_path, entry)
+            for entry in os.listdir(local_path)
+            if entry.startswith("checkpoint_")
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/result.py:158]: Good point! 
Should we move this into some where in `ray/air`, otherwise we will run into cyclic import issue (AIR->Tune->AIR) again?

File: python/ray/air/result.py [hunk 11]
```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
...
+        ]
+
+        if ckpt_dirs:
+            checkpoints = [
+                Checkpoint.from_directory(ckpt_dir) for ckpt_dir in ckpt_dirs
+            ]
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/result.py:164]: hmm that's a problem. Is there anyway we can get the checkpoint S3 uri from the artifact in trial folder?

File: python/ray/air/result.py [hunk 12]
```diff
@@ -110,3 +117,109 @@ def _repr(self, indent: int = 0) -> str:
...
+
+            checkpoint_metrics = [
+                metrics_df[metrics_df[TRAINING_ITERATION] == ckpt.iteration].to_dict(
+                    "records"
+                )[0]
+                for ckpt in checkpoints
+            ]
+
+            # TODO(air-team): make metrics a property of Checkpoint
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/result.py:173]: Yeah, that's what I am thinking of! Dealing with a list of tuples (ckpt, metric) is so weird for me.

File: python/ray/air/result.py (hunk 1 above)
Comment (woshiyyya) [reply python/ray/air/result.py:182]: ok

File: python/ray/air/result.py (hunk 2 above)
Comment (woshiyyya) [reply python/ray/air/result.py:196]: I think this should be considered together with making metrics a property of Checkpoint. 

For example, saving a json file in the checkpoint folder with something like:

```
{"checkpoint_score_attribute": "abc", "checkpoint_score_order": "max", "latest_metrics": {"abc": 123, "xyz": 456}}
```

I think we can open a separate PR for this, since it introduces more code changes.

File: python/ray/air/result.py (hunk 4 above)
Comment (woshiyyya) [reply python/ray/air/result.py:197]: make sense

File: python/ray/air/result.py (hunk 3 above)
Comment (woshiyyya) [reply python/ray/air/result.py:219]: If no valid checkpoints, we can only return None? so it has to be non-optional?

**Comment by woshiyyya** [comment]
@justinvyu To summarize, these are the improvements for the following PRs:
- Figure out a way to restore result without checkpoint dirs (cloud based checkpoint)
- Make metric and mode as properties of `Checkpoint`
- Have an util function to check if a directory is a checkpoint directory.

File: python/ray/air/checkpoint.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/checkpoint.py]: OK, I removed those properties. I am extracting ckpt iteration from `.tune_metadata` now.

File: python/ray/air/tests/test_result.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    elif mode == "tuner":
+        tuner = build_dummy_tuner(configs)
+        tuner.fit()
+
+    # Find the trial directory to restore
+    exp_dir = os.path.join(storage_path, exp_name)
+    for dirname in os.listdir(exp_dir):
+        if dirname.startswith("TorchTrainer"):
+            trial_dir = os.path.join(exp_dir, dirname)
+            break
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/air/tests/test_result.py:78]: Hmm, ideally we don't need to run Tune/Train and we could just mock the trial directory using the default loggers. But the Tune metadata thing gets created by Tune Trainable.save, which is a bit harder to mock.

I think this is ok for now, but wondering how we can improve testing to not do a full run each time. There's [this testing harness](https://github.com/ray-project/ray/blob/7c7361f93741e57149d571c4bdfe76c3e74c7953/python/ray/tune/tests/utils/experiment.py#L103) that kind of does what we want, but not exactly because it doesn't create real trial checkpoints.

File: python/ray/air/result.py [hunk 14]
```diff
@@ -48,6 +58,7 @@ class Result:
     _local_path: Optional[str] = None
     _remote_path: Optional[str] = None
     _items_to_repr = ["error", "metrics", "path", "checkpoint"]
+    _restore_required_files = [EXPR_RESULT_FILE]
```
Comment (justinvyu) [inline python/ray/air/result.py:61]: Oh, for this, I meant:

- We shouldn't raise a warning if the user doesn't have *both* files. They could have run with only the JSON logger or only the CSV logger. The user must run with the JSON logger for the current code to work.
- What do you think about:
    - Fallback to CSV if JSON not found.
    - Just keep it as is and require JSON, I think that's fine since 99% of the time users will run with both (that's the default). But, we should plan when to get rid of csv logger.

**Review by justinvyu — COMMENTED** [review]
Thanks for addressing the previous round. Some minor comments:

File: python/ray/air/result.py (hunk 3 above)
Comment (woshiyyya) [reply python/ray/air/result.py:242]: Got it, that makes a lot of sense.

File: python/ray/air/tests/test_result.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/tests/test_result.py:78]: Yeah I was trying to use `result.path` at the beginning, but when we throw an error in the `worker_loop()`, the trainer fails before returning a result object, That's why I use `dirname.startswith("TorchTrainer")` to find the trial dir.

Yeah I agree, currently we have to start e2e training/tuning to generate a trial folder, which conflict with the idea of unit test. It'd be better to have a `generate_trial_folder()` mock function, which aligns with the behavior of real trainer & tuner.

File: python/ray/air/tests/test_result.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    # Check if we restored all checkpoints
+    assert result.checkpoint
+    assert len(result.best_checkpoints) == NUM_CHECKPOINTS
+
+    # Check if the checkpoints bounded with correct metrics
+    best_ckpt_a = result.get_best_checkpoint(metric="metric_a", mode="max")
+    assert best_ckpt_a.to_dict()["iter"] == NUM_ITERATIONS - 1
+
+    best_ckpt_b = result.get_best_checkpoint(metric="metric_b", mode="max")
+    assert best_ckpt_b.to_dict()["iter"] == NUM_ITERATIONS - NUM_CHECKPOINTS
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/tests/test_result.py:92]: yeah sure.

File: python/ray/air/result.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/air/result.py:177]: Ok

File: python/ray/air/result.py (hunk 14 above)
Comment (woshiyyya) [reply python/ray/air/result.py:61]: Oh got it. In that case, I prefer to fallback to loading CSV if there is no JSON for robustness. Then raise a deprecation warning if they are using csv logger.

**Review by krfricke — APPROVED** [review]
Thanks! Looks great!
